pkg/clusteragent/admission/mutate/autoinstrumentation/injector.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2016-present Datadog, Inc.
+
+//go:build kubeapiserver
+
+package autoinstrumentation
+
+import (
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+var v1VolumeMount = volumeMount{
+	mount: corev1.VolumeMount{
+		Name:      volumeName,
+		MountPath: mountPath,
+	},
+}
+
+var v2VolumeMountInjector = corev1.VolumeMount{
+	Name:      volumeName,
+	MountPath: "/opt/datadog-packages/datadog-apm-inject",
+	SubPath:   "opt/datadog-packages/datadog-apm-inject",
+}
+
+var v2VolumeMountLibrary = corev1.VolumeMount{
+	Name:      volumeName,
+	MountPath: "/opt/datadog/apm/library",
+	SubPath:   "opt/datadog/apm/library",
+}
+
+var sourceVolume = volume{
+	volume: corev1.Volume{
+		Name: volumeName,
+		VolumeSource: corev1.VolumeSource{
+			EmptyDir: &corev1.EmptyDirVolumeSource{},
+		},
+	},
+}
+
+type injector struct {
+	image    string
+	registry string
+	injected bool
+}
+
+func (i *injector) initContainer() initContainer {
+	mount := v2VolumeMountInjector
+	mount.MountPath = "/datadog-inject"
+	return initContainer{
+		prepend: true,
+		container: corev1.Container{
+			Name:         "dd-init-apm-inject",
+			Image:        i.image,
+			Command:      []string{"/bin/sh", "-c", "--"},
+			Args:         []string{fmt.Sprintf("cp -r /%s/* %s", mount.SubPath, mount.MountPath)},
+			VolumeMounts: []corev1.VolumeMount{mount},
+		},
+	}
+}
+
+func (i *injector) requirements() libRequirement {
+	// we mount RO volumes
+	mount := v2VolumeMountInjector
+	mount.ReadOnly = true
+
+	return libRequirement{
+		initContainers: []initContainer{i.initContainer()},
+		volumes:        []volume{sourceVolume},
+		volumeMounts:   []volumeMount{{mount: mount, prepend: true}},
+		envVars: []envVar{
+			{
+				key: "LD_PRELOAD",
+				valFunc: identityValFunc(
+					"/opt/datadog-packages/datadog-apm-inject/stable/inject/launcher.preload.so",
+				),
+			},
+			{
+				key: "DD_INJECT_SENDER_TYPE",
+				valFunc: identityValFunc(
+					"k8s",
+				),
+			},
+		},
+	}
+}
+
+type injectorOption func(*injector)
+
+var injectorVersionAnnotationExtractor = annotationExtractor[injectorOption]{
+	key: "admission.datadoghq.com/apm-injector.version",
+	do:  infallibleFn(injectorWithImageTag),
+}
+
+var injectorImageAnnotationExtractor = annotationExtractor[injectorOption]{
+	key: "admission.datadoghq.com/apm-injector.custom-image",
+	do:  infallibleFn(injectorWithImageName),
+}
+
+func injectorWithImageName(name string) injectorOption {
+	return func(i *injector) {
+		i.image = name
+	}
+}
+
+func injectorWithImageTag(tag string) injectorOption {
+	return func(i *injector) {
+		i.image = fmt.Sprintf("%s/apm-injector:%s", i.registry, tag)
+	}
+}
+
+func newInjector(registry string, pod *corev1.Pod) *injector {
+	i := &injector{
+		registry: registry,
+	}
+
+	for _, e := range []annotationExtractor[injectorOption]{
+		injectorVersionAnnotationExtractor,
+		injectorImageAnnotationExtractor,
+	} {
+		if opt, ok, _ := e.extract(pod); ok && opt != nil {
+			opt(i)
+		}
+	}
+
```

Comment:
Implementing custom image/tag loading for the injector package using annotationExtractor.